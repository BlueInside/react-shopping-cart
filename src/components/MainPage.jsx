import { useNavigate } from 'react-router-dom';
import Button from '../components/Button';
import useDataFetching from '../hooks/useDataFetching';

function MainPage() {
  const navigate = useNavigate();

  const { data, error, loading } = useDataFetching();

  if (loading || !data) return <p role="loader">Loading...</p>;
  if (error) return <p role="error">Error</p>;
  return (
    <main>
      <h1>Trend Tribe: Join the Fashion Revolution</h1>
      <div>
        {data &&
          data.slice(0, 3).map((item) => {
            return <img key={item.id} src={item.image} />;
          })}
      </div>
      <Button handleClick={() => navigate('/shop')} label={'Discover more'} />
    </main>
  );
}

export default MainPage;
