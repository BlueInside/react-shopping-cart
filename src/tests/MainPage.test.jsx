import { render, screen, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import MainPage from '../components/MainPage';
import { MemoryRouter } from 'react-router-dom';

describe('MainPage', () => {
  const renderWithRouter = (ui, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);

    return render(ui, { wrapper: MemoryRouter });
  };

  it('renders MainPage component correctly', () => {
    const { container } = renderWithRouter(<MainPage />);
    expect(container).toMatchSnapshot();
  });

  it('should have header "Trend Tribe: Join the Fashion Revolution"', async () => {
    renderWithRouter(<MainPage />);

    const header = await screen.findByRole('heading', {
      name: /Trend Tribe: Join the Fashion Revolution/i,
    });
    expect(header).toBeInTheDocument();
  });

  it('contains, "discover more" button', async () => {
    renderWithRouter(<MainPage />);
    const button = await screen.findByRole('button', {
      label: /Discover more/i,
    });
    expect(button).toBeInTheDocument();
  });

  it('displays loading paragraph', () => {
    renderWithRouter(<MainPage />);
    const loadingPara = screen.getByRole('loader', { value: /Loading.../i });
    expect(loadingPara).toBeInTheDocument();
  });

  it('should fetch and display 3 images', async () => {
    renderWithRouter(<MainPage />);
    await waitFor(() => {
      const images = screen.getAllByRole('img');
      expect(images).toHaveLength(3);
    });
  });

  it('displays error paragraph', async () => {
    globalThis.fetch = vi.fn(async () => {
      Promise.resolve({
        ok: false,
        json: () => {
          return Promise.reject(new Error('Failed to fetch data'));
        },
      });

      renderWithRouter(<MainPage />);

      const errorPara = screen.getByRole('error', {
        value: /Failed to fetch data/i,
      });

      expect(errorPara).toBeInTheDocument();
    });
  });
});
