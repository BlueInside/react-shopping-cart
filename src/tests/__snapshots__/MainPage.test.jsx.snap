// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainPage > renders MainPage component correctly 1`] = `
<div>
  <p
    role="loader"
  >
    Loading...
  </p>
</div>
`;
